### PR TITLE
Add pass/fail convergence test for ocean test cases 

### LINKED ISSUE #230: Add pass/fail test for convergence tests
The wealth of sphere_transport tests introduced in #210 would benefit with some form of pass/fail tests.  This would enable easier integration into nightly suites and would allow for no room in interpretation as to if a test succeeded.

### PR DESCRIPTION
**Pull request by yariseidenbenz**
This merge adds a convergence threshold to mesh convergence test in the ocean. Previously mesh convergence tests ran MPAS-ocean with multiply mesh resolutions but did not check the order of convergence. The changes in this branch raise an exception when the convergence order is slower or faster than expected.  The tests that are modified are:
* sphere_transport/rotation_2d
* sphere_transport/nondivergent_2d
* sphere_transport/divergent_2d
* sphere_transport/correlated_tracers_2d
* global_convergence/cosine_bell
  
closes #230

### CODE REVIEW

File: compass/ocean/tests/sphere_transport/divergent_2d/analysis.py [hunk 1]
```diff
@@ -104,3 +104,15 @@ def run(self):
...
+        conv_thresh = section.getfloat('conv_thresh')
+        conv_max = section.getfloat('conv_max')
+
+        if conv < conv_thresh:
+            raise ValueError(f'order of convergence '
+                             f'{conv} < min tolerence {conv_thresh}')
+
+        if conv > conv_max:
+            warnings.warn(f'order of convergence '
+                          f'{conv} > max tolerence {conv_max}')
```
[the remark this answers is not in the document]
Comment (xylar) [reply compass/ocean/tests/sphere_transport/divergent_2d/analysis.py:118]: (A blank line at the end of the file)

File: compass/ocean/tests/sphere_transport/nondivergent_2d/analysis.py [hunk 2]
```diff
@@ -142,3 +142,16 @@ def run(self):
...
+        conv_max = section.getfloat('conv_max')
+
+        if conv < conv_thresh:
+            raise ValueError(f'order of convergence '
+                             f'{conv} < min tolerence {conv_thresh}')
+
+        if conv > conv_max:
+            warnings.warn(f'order of convergence '
+                          f'{conv} > max tolerence {conv_max}')
+
```
[the remark this answers is not in the document]
Comment (xylar) [reply compass/ocean/tests/sphere_transport/nondivergent_2d/analysis.py:157]: Just one, not two, blank lines at the end of the file

**Comment by xylar** [comment]
Excellent work so far, @yariseidenbenz.  I did a test run with this branch and I discovered that the various `sphere_transport` tests turn out to be more complicated than I expected.  The complications are 2:
* Each test involves 3 fields, not just one
* The individual dots on the graph are computed but not the line passing through them
So getting those 4 tests working will involve some more code.

We can return to this once I'm back in Germany and have a little time.  If @matthewhoffman has other, more pressing projects, I can also make some headway here and then we can finish up this project together.

**Comment by xylar** [comment]
One way forward, and I think the best one, is that I figure out what's needed for one of the `sphere_transport` test cases and @yariseidenbenz can then do the same for the other 3.  How does that sound?

**Comment by matthewhoffman** [comment]
@xylar , thanks for the update.  We have other stuff we can work on in the meantime, and we can return to this when you could use more help from @yariseidenbenz .

**Comment by xylar** [comment]
@yariseidenbenz, this is great!

I think the in the last file isn't quite right, though, and it's causing the tests to fail.  Maybe try copying and pasting that code again and see if it matches the others?  If so, you can do:
```
git commit --amend compass/ocean/tests/sphere_transport/rotation_2d/analysis.py
git push --force yariseidenbenz/compass add_pass_fail_converge
```
To edit your last commit and then update things here.

**Comment by xylar** [comment]
## Notes from today's work

@yariseidenbenz and I tested on Badger (with intel and impi) and Chrysalis (same compilers and MPI).  Things "worked" on Chrysalis in the sense that we got the expected failures:
```
rotation_2d:
ValueError: The following tracers have order of convergence < min tolerance:
            tracer1: 1.46 < 5.0
            tracer2: 1.85 < 5.0
            tracer3: 0.43 < 4.0

nondivergent_2d:
ValueError: The following tracers have order of convergence < min tolerance:
            tracer1: 1.77 < 5.0
            tracer2: 1.76 < 5.0
            tracer3: 0.41 < 4.0

divergent_2d:
ValueError: The following tracers have order of convergence < min tolerance:
            tracer1: 1.87 < 5.0
            tracer2: 2.00 < 5.0
            tracer3: 0.42 < 4.0
```
But there were some MPI-related error son Badger that I need to investigate more thoroughly. It may be that we needed to rebuild the conda environment that @yariseidenbenz is using but I can't think off the top of my head of a reason that would be the case.

We also saw some debug output that should be removed:
```
[0.00132604 0.00262601 0.00461511 0.00702342 0.01020573 0.01716098]
[165004  73453  41376  26507  18430  10383]
tracer1
1.8712086549952867
[0.01841266 0.04126311 0.0804899  0.12506597 0.18440102 0.27263066]
[165004  73453  41376  26507  18430  10383]
tracer2
2.003410798090883
[0.25184152 0.30365774 0.35411403 0.388628   0.41561213 0.44158992]
[165004  73453  41376  26507  18430  10383]
tracer3
0.42118097979316677
```
I'll track that down and flag it for removal

File: compass/ocean/tests/sphere_transport/correlated_tracers_2d/analysis.py [hunk 3]
```diff
@@ -115,4 +115,36 @@ def run(self):
...
+            conv_thresh = section.getfloat(f'{tracer}_conv_thresh')
+            l2_err = list()
+            ncells = list()
+            for resolution in self.resolutions:
+                data = self.tcdata[resolution]
+                l2_err.append(data['err'][tracer]['l2'])
+                ncells.append(len(data['dataset'].dimensions["nCells"]))
+            l2_err = np.array(l2_err)
+            ncells = np.array(ncells)
+            print(l2_err, ncells)
```
Comment (xylar) [inline compass/ocean/tests/sphere_transport/correlated_tracers_2d/analysis.py:134]: Should be removed.
```suggestion
```

File: compass/ocean/tests/sphere_transport/correlated_tracers_2d/analysis.py [hunk 4]
```diff
@@ -115,4 +115,36 @@ def run(self):
...
+            p = np.polyfit(np.log10(ncells), np.log10(l2_err), 1)
+
+            # factor of 2 because nCells is like an inverse area, and we
+            # want the convergence rate vs. cell size
+            conv = abs(p[0]) * 2.0
+            print(tracer, conv)
```
Comment (xylar) [inline compass/ocean/tests/sphere_transport/correlated_tracers_2d/analysis.py:140]: Also should be removed
```suggestion
```

File: compass/ocean/tests/sphere_transport/correlated_tracers_2d/analysis.py (hunk 3 above)
Comment (xylar) [reply compass/ocean/tests/sphere_transport/correlated_tracers_2d/analysis.py:134]: The same in each of the `analysis` steps.

**Comment by xylar** [comment]
I ran a test on Badger myself and all 4 test cases in `sphere_transport` passed this time.  I think there might be something that got a little messed up in @yariseidenbenz's configuration.  Maybe something related to working with MALI.  Presumably, rebuilding the conda environment and MPAS-Ocean would take care of it.

**Comment by xylar** [comment]
Thanks for the hard work on this @yariseidenbenz!  I'm going to go ahead and merge it once the automated tests pass.

**Comment by xylar** [comment]
@vanroekel, I just wanted to note that I set convergence thresholds for each tracer in the spherical transport tests so that they barely pass.  

As we saw in #210, tracer 3, which is c-0 discontinuous, has very poor convergence properties.

Presumably, if changes in numerical methods improve convergence, we should increase these thresholds accordingly.

**Pull request merged**